### ISSUE TITLE
Segmentation fault when calling InitVulkan and passing an AMFVulkanDevice

### ISSUE THREAD

**Issue by mfleisz** [GPUOpen-LibrariesAndSDKs/AMF#145]
Whenever I try to pass an AMFVulkanDevice into AMFContext1::InitVulkan I'm getting a segmentation fault:

```
amf::AMFVulkanDevice vulkan_dev;

vulkan_dev.cbSizeof = sizeof(vulkan_dev);
vulkan_dev.pNext = nullptr;
vulkan_dev.hDevice = device->device;
vulkan_dev.hInstance = device->instance;
vulkan_dev.hPhysicalDevice = device->phys_device;
res = amf::AMFContext1Ptr(amf_context)->InitVulkan(&vulkan_dev);  // this segfaults
```

Here is the callstack:
```
[Unknown/Just-In-Time compiled code] (Unknown Source:0)
libamfrt64.so.1!amf::AMFDeviceVulkanImpl::CreateDeviceAndFindQueues(bool) (Unknown Source:0)
libamfrt64.so.1!amf::AMFDeviceVulkanImpl::Init(void*) (Unknown Source:0)
libamfrt64.so.1!AMFContextImpl::InitVulkan(void*) (Unknown Source:0)
vulkan_v3d.so!amf_video_encoder::init(v3d10_amf_video_encoder * const this, v3d_context * context, uint32_t w, uint32_t h, void * evt, int flags) (----removed-----)
vulkan_v3d.so!video_create_encoder(v3d_context * context, unsigned int width, unsigned int height, void * evt, int flags) (------removed-------)
......
libpthread.so.0!start_thread(void * arg) (/build/glibc-OTsEL5/glibc-2.27/nptl/pthread_create.c:463)
libc.so.6!clone() (/build/glibc-OTsEL5/glibc-2.27/sysdeps/unix/sysv/linux/x86_64/clone.S:95)
```

When passing nullptr to InitVulkan the function succeeds with AMF_OK.
The passed device was successfully created (including the extensions returned by AMFContext1::GetVulkanDeviceExtensions).
The documentation isn't of much help so I'm kinda lost what I might be doing wrong ... my only guess is that it must be something during instance/device creation ...?

I'm on Ubuntu 18.04.1 LTS with all updates and latest amdgpu-pro drivers installed.

**Comment by MikhailAMD** [comment]
Did you try VCEEncoderD3D sample? It also creates device outside of AMF.

**Comment by mfleisz** [comment]
@MikhailAMD Thanks for pointing me to the sample. The sample works fine! I'll try to figure out what it does different than my code and post my findings here! (AMFContext1::InitVulkan should not segfault due to invalid args)

**Comment by MikhailAMD** [comment]
Agree with you on crash point. Please post the issue when you find it. We will use all information available to improve.

**Comment by mfleisz** [comment]
@MikhailAMD I'm still trying to figure out what's wrong in my project but I found a way to cause a very similar segfault using the VCEEncoderD3D sample.
On my Radeon RX 480 ```vkGetPhysicalDeviceQueueFamilyProperties``` reports 6 queue families and the ```DeviceVulkan``` class will create queues for all 6 of them. If I only create queues for the first 2 families (by setting ```queueFamilyPropertyCount``` to 2 - these queues support graphics and compute) I receive a segfault very similar to the one in my project.

I looked in the documentation but it mentions nowhere what requirements InitVulkan has on the passed device. That information would be very helpful. Also I think it should be mentioned somewhere in the docs that you have to init AMF *before* creating your vkInstance or you won't be able to create a device that has all the AMD AMF extensions enabled ... (but that's a different topic I guess)

I'll debug my project further and see if I find something else that's causing a segfault (it still segfaults even though I'm initializing all available queues but I might miss something else in my code ...).

**Comment by mfleisz** [comment]
I also noted that the ```cbSizeof``` member of the ```AMFVulkanDevice``` structure is not set correctly in the samples and the runtime apparently doesn't check if it is set to a valid size.

**Comment by llyzs** [comment]
So far I found two causes that could crash the InitVulkan call.

1. The first crash point is because AMF is not compatible the validation layer. If the Vulkan instance is created with validation layer enabled, the AMF driver will make a null pointer function call, somewhere at vkPhysDevExtTramp19

1. The Vulkan device must create some specific queues before it can be used by InitVulkan. On RX480 I found that it is queue family #3 and #4. Missing any of the two queue families during Vulkan device creation will cause AMF driver to crash. I think AMF driver should report meaningful error for this behavior.
